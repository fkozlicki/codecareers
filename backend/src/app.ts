import bodyParser from 'body-parser';
import cors from 'cors';
import dotenv from 'dotenv';
import express, { Request, Response } from 'express';
import { corsOptions } from './config/corsOptions';
import { serveFile } from './lib/s3';
import { authorizeCv } from './middleware/authorization';
import { requireSession, verifySession } from './middleware/session';
import { applicationsRouter } from './routes/application.routes';
import { authRouter } from './routes/auth.routes';
import { chatsRouter } from './routes/chat.routes';
import { companiesRouter } from './routes/company.routes';
import { jobOffersRouter } from './routes/jobOffer.routes';
import { recruitmentsRouter } from './routes/recruitment.routes';
import { skillsRouter } from './routes/skill.routes';
import { technologiesRouter } from './routes/technology.routes';
import { usersRouter } from './routes/user.routes';

dotenv.config({ path: `.env.${process.env.NODE_ENV ?? 'development'}` });

const app = express();

app.use(bodyParser.json());
app.use(bodyParser.urlencoded({ extended: true }));
app.use(cors(corsOptions));

app.use(verifySession);

app.get('/', (req: Request, res: Response) => {
	res.json({ message: 'Welcome to CodeCareers API' });
});

app.use('/', authRouter);
app.use('/companies', companiesRouter);
app.use('/job-offers', jobOffersRouter);
app.use('/technologies', technologiesRouter);
app.use('/skills', skillsRouter);
app.use('/applications', applicationsRouter);
app.use('/users', usersRouter);
app.use('/recruitments', recruitmentsRouter);
app.use('/chats', chatsRouter);
app.use('/avatars/:filename', serveFile('avatars'));
app.use('/cv/:filename', requireSession, authorizeCv, serveFile('cvs'));

export default app;
