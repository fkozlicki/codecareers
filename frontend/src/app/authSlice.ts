import { User, session } from '@/app/services/auth';
import { PayloadAction, createSlice } from '@reduxjs/toolkit';

interface AuthState {
	user: User | null;
	status: 'loading' | 'authenticated' | 'unauthenticated';
}

const initialState: AuthState = {
	user: null,
	status: 'loading',
};

const { reducer, actions } = createSlice({
	name: 'auth',
	initialState,
	reducers: {
		logout: () => initialState,
		setUser: (state, action: PayloadAction<User>) => {
			state.user = action.payload;
		},
	},
	extraReducers: (builder) => {
		builder
			.addMatcher(session.matchPending, (state) => {
				state.status = 'loading';
			})
			.addMatcher(session.matchFulfilled, (state, action) => {
				if (action.payload.user) {
					state.status = 'authenticated';
					state.user = action.payload.user;
				} else {
					state.status = 'unauthenticated';
				}
			});
	},
});

export const { logout, setUser } = actions;

export default reducer;
