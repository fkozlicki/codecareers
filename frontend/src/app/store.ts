import auth from '@/app/authSlice';
import { api } from '@/app/services/api';
import { combineReducers, configureStore } from '@reduxjs/toolkit';
import { setupListeners } from '@reduxjs/toolkit/query';

const rootReducer = combineReducers({
	[api.reducerPath]: api.reducer,
	auth,
});

export const setupStore = (preloadedState?: Partial<RootState>) => {
	return configureStore({
		reducer: rootReducer,
		middleware: (getDefaultMiddleware) =>
			getDefaultMiddleware().concat(api.middleware),
		preloadedState,
	});
};

export const store = setupStore();

setupListeners(store.dispatch);

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof rootReducer>;
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch;
export type AppStore = ReturnType<typeof setupStore>;
