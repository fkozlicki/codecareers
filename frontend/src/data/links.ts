import {
	Building2Icon,
	FilePenIcon,
	LogOutIcon,
	SettingsIcon,
	UsersIcon,
} from 'lucide-react';

export const navigationLinks = [
	{
		label: 'Applications',
		Icon: FilePenIcon,
		href: '/my-applications',
	},
	{
		label: 'Recruitments',
		Icon: UsersIcon,
		href: '/my-recruitments',
	},
	{
		label: 'Companies',
		Icon: Building2Icon,
		href: '/my-companies',
	},
	{
		label: 'Settings',
		Icon: SettingsIcon,
		href: '/settings',
	},
	{
		label: 'Logout',
		Icon: LogOutIcon,
		href: `${import.meta.env.VITE_API_URI}/logout`,
	},
];
