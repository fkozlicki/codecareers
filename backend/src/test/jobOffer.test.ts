import supertest from 'supertest';
import { TestDatabase, cleanupDB, initializeDB, password } from './data.js';
import app from '../app.js';

let db: TestDatabase;
let adamSession = '';
let jonSession = '';

beforeEach(async () => {
	const database = await initializeDB();
	db = database;

	// sign in
	const adamResult = await supertest(app).post('/login/credentials').send({
		email: database.adam.email,
		password,
	});
	adamSession = adamResult.header['set-cookie'];

	const jonResult = await supertest(app).post('/login/credentials').send({
		email: database.jon.email,
		password,
	});

	jonSession = jonResult.header['set-cookie'];
});

afterEach(async () => {
	await cleanupDB();
});

describe('JobOffer service', () => {
	describe('[GET] /job-offers', () => {
		it('Should return list of public job-offers', async () => {
			const { statusCode, body } = await supertest(app).get('/job-offers');

			expect(statusCode).toBe(200);
			for (const jobOffer of body.jobOffers) {
				expect(jobOffer).toMatchSnapshot({
					id: expect.any(String),
					companyId: expect.any(String),
					createdAt: expect.any(String),
					company: { id: expect.any(String), ownerId: expect.any(String) },
				});
			}
		});
	});

	describe('[GET] /job-offers/:id', () => {
		it('Should return job offer', async () => {
			const { statusCode, body } = await supertest(app).get(
				`/job-offers/${db.seniorWebDeveloper.id}`
			);

			expect(statusCode).toBe(200);
			expect(body.jobOffer).toMatchSnapshot({
				id: expect.any(String),
				createdAt: expect.any(String),
				companyId: expect.any(String),
				company: {
					id: expect.any(String),
					ownerId: expect.any(String),
				},
			});
		});
	});

	describe('[PUT] /job-offers/:id', () => {
		it('Should update job offer', async () => {
			const { statusCode, body } = await supertest(app)
				.put(`/job-offers/${db.seniorWebDeveloper.id}`)
				.set('cookie', adamSession)
				.send({
					position: 'Junior Web Developer',
					description: 'Looking for passionate junior web developer',
					level: 'junior',
				});

			expect(statusCode).toBe(200);
			expect(body.jobOffer).toMatchSnapshot({
				id: expect.any(String),
				createdAt: expect.any(String),
				companyId: expect.any(String),
			});
		});
	});

	describe('[POST] /job-offers/:id/applications', () => {
		it('Should create application for job offer', async () => {
			const { statusCode, body } = await supertest(app)
				.post(`/job-offers/${db.seniorWebDeveloper.id}/applications`)
				.set('cookie', jonSession)
				.send({
					introduction:
						'Hello Google, I would love to work for you as Senior Web Developer',
				});

			expect(statusCode).toBe(201);
			expect(body.application).toMatchSnapshot({
				id: expect.any(String),
				createdAt: expect.any(String),
				jobOfferId: expect.any(String),
				userId: expect.any(String),
			});
		});
	});

	describe('[GET] /job-offers/:id/applications', () => {
		it('Should return list of applications for job offer', async () => {
			const { statusCode, body } = await supertest(app)
				.get(`/job-offers/${db.seniorWebDeveloper.id}/applications`)
				.set('cookie', adamSession);

			expect(statusCode).toBe(200);
			for (const application of body.applications) {
				expect(application).toMatchSnapshot({
					id: expect.any(String),
					jobOfferId: expect.any(String),
					createdAt: expect.any(String),
					user: {
						id: expect.any(String),
					},
				});
			}
		});
	});
});
