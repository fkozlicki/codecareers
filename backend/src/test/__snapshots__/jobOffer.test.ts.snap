// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JobOffer service [GET] /job-offers Should return list of public job-offers 1`] = `
{
  "company": {
    "avatarUrl": null,
    "backgroundUrl": null,
    "description": null,
    "id": Any<String>,
    "name": "Google",
    "ownerId": Any<String>,
    "phoneNumber": "222 222 2222",
  },
  "companyId": Any<String>,
  "createdAt": Any<String>,
  "description": "",
  "employmentType": "b2b",
  "id": Any<String>,
  "level": "senior",
  "position": "Senior Web Developer",
  "published": true,
  "salaryCurrency": "pln",
  "salaryFrom": 25000,
  "salaryTo": 30000,
  "workType": "full_time",
}
`;

exports[`JobOffer service [GET] /job-offers/:id Should return job offer 1`] = `
{
  "company": {
    "avatarUrl": null,
    "backgroundUrl": null,
    "description": null,
    "id": Any<String>,
    "name": "Google",
    "ownerId": Any<String>,
    "phoneNumber": "222 222 2222",
  },
  "companyId": Any<String>,
  "createdAt": Any<String>,
  "description": "",
  "employmentType": "b2b",
  "id": Any<String>,
  "jobOfferSkills": [],
  "jobOfferTechnologies": [],
  "level": "senior",
  "position": "Senior Web Developer",
  "published": true,
  "salaryCurrency": "pln",
  "salaryFrom": 25000,
  "salaryTo": 30000,
  "workType": "full_time",
}
`;

exports[`JobOffer service [GET] /job-offers/:id/applications Should return list of applications for job offer 1`] = `
{
  "accepted": null,
  "createdAt": Any<String>,
  "cv": null,
  "id": Any<String>,
  "introduction": "Hello Google, I want to work for you",
  "jobOfferId": Any<String>,
  "user": {
    "avatar": null,
    "email": "jon.snow@gmail.com",
    "firstName": "Jon",
    "githubId": null,
    "id": Any<String>,
    "lastName": "Snow",
    "username": null,
  },
}
`;

exports[`JobOffer service [POST] /job-offers/:id/applications Should create application for job offer 1`] = `
{
  "accepted": null,
  "createdAt": Any<String>,
  "cv": null,
  "id": Any<String>,
  "introduction": "Hello Google, I would love to work for you as Senior Web Developer",
  "jobOfferId": Any<String>,
  "userId": Any<String>,
}
`;

exports[`JobOffer service [PUT] /job-offers/:id Should update job offer 1`] = `
{
  "companyId": Any<String>,
  "createdAt": Any<String>,
  "description": "Looking for passionate junior web developer",
  "employmentType": "b2b",
  "id": Any<String>,
  "level": "junior",
  "position": "Junior Web Developer",
  "published": true,
  "salaryCurrency": "pln",
  "salaryFrom": 25000,
  "salaryTo": 30000,
  "workType": "full_time",
}
`;
